### ISSUE TITLE
TypeError: Cannot read property 'query' of null

### ISSUE THREAD

**Issue by deebloo** [angular/angular#8506]
I am having issues with RC1 router-deprecated with webpack. Apologies if this is answered somewhere else but I could not find it.

Error:
```
EXCEPTION: Error: Uncaught (in promise): TypeError: Cannot read property 'query' of null
browser_adapter.js:77EXCEPTION: Error: Uncaught (in promise): TypeError: Cannot read property 'query' of nullBrowserDomAdapter.logError @ browser_adapter.js:77BrowserDomAdapter.logGroup @ browser_adapter.js:87ExceptionHandler.call @ exception_handler.js:57(anonymous function) @ application_ref.js:265schedulerFn @ async.js:123SafeSubscriber.__tryOrUnsub @ Subscriber.js:225SafeSubscriber.next @ Subscriber.js:174Subscriber._next @ Subscriber.js:124Subscriber.next @ Subscriber.js:88Subject._finalNext @ Subject.js:128Subject._next @ Subject.js:120Subject.next @ Subject.js:77EventEmitter.emit @ async.js:112NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.js:120NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.js:66ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494
browser_adapter.js:77STACKTRACE:BrowserDomAdapter.logError @ browser_adapter.js:77ExceptionHandler.call @ exception_handler.js:59(anonymous function) @ application_ref.js:265schedulerFn @ async.js:123SafeSubscriber.__tryOrUnsub @ Subscriber.js:225SafeSubscriber.next @ Subscriber.js:174Subscriber._next @ Subscriber.js:124Subscriber.next @ Subscriber.js:88Subject._finalNext @ Subject.js:128Subject._next @ Subject.js:120Subject.next @ Subject.js:77EventEmitter.emit @ async.js:112NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.js:120NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.js:66ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494
browser_adapter.js:77Error: Uncaught (in promise): TypeError: Cannot read property 'query' of null
    at resolvePromise (zone.js:538)
    at zone.js:515
    at ZoneDelegate.invoke (zone.js:323)
    at Object.NgZoneImpl.inner.inner.fork.onInvoke (ng_zone_impl.js:45)
    at ZoneDelegate.invoke (zone.js:322)
    at Zone.run (zone.js:216)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Object.NgZoneImpl.inner.inner.fork.onInvokeTask (ng_zone_impl.js:36)
    at ZoneDelegate.invokeTask (zone.js:355)BrowserDomAdapter.logError @ browser_adapter.js:77ExceptionHandler.call @ exception_handler.js:60(anonymous function) @ application_ref.js:265schedulerFn @ async.js:123SafeSubscriber.__tryOrUnsub @ Subscriber.js:225SafeSubscriber.next @ Subscriber.js:174Subscriber._next @ Subscriber.js:124Subscriber.next @ Subscriber.js:88Subject._finalNext @ Subject.js:128Subject._next @ Subject.js:120Subject.next @ Subject.js:77EventEmitter.emit @ async.js:112NgZone._zoneImpl.ng_zone_impl_1.NgZoneImpl.onError @ ng_zone.js:120NgZoneImpl.inner.inner.fork.onHandleError @ ng_zone_impl.js:66ZoneDelegate.handleError @ zone.js:327Zone.runGuarded @ zone.js:233_loop_1 @ zone.js:487drainMicroTaskQueue @ zone.js:494
zone.js:461 Unhandled Promise rejection: Cannot read property 'query' of null ; Zone: angular ; Task: Promise.then ; Value: TypeError: Cannot read property 'query' of null(…)consoleError @ zone.js:461_loop_1 @ zone.js:490drainMicroTaskQueue @ zone.js:494
zone.js:463 Error: Uncaught (in promise): TypeError: Cannot read property 'query' of null(…)
```

webpack.config.js
```JS
module.exports = {  
  entry: './src/main.ts',
  output: {
    filename: 'bundle.js',
  },
  devtool: 'source-map',
  resolve: {
    extensions: ['', '.webpack.js', '.web.js', '.tsx', '.ts', '.js']
  },
  module: {
    loaders: [
      { test: /\.tsx|ts$/, loader: 'ts-loader' }
    ]
  }
}
```

tsconfig.json
```JSON
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "es5",
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true
    },
    "exclude": [
        "node_modules",
        "typings/main",
        "typings/main.d.ts"
    ]
}
```

**Comment by zoechi** [comment]
Seems you have a wrong or missing import. Can you reproduce in a Plunker?

**Comment by deebloo** [comment]
I will try. It is just odd because it works fine with systemjs and module type of 'system' but craps out with webpack and a module type of 'commonjs'

**Comment by deebloo** [comment]
so it looks like it is ending up being something wrong with my barrel. for some reason when I am compiling with webpack it seems to care a lot about what order I import the different modules. :/ i still can't put my finger on what is happeneing.

**Comment by deebloo** [comment]
im closing this issue since it doesn't seem to have anything to do with angular specifically. I would say it is more to do with crypt unhelpful error messages from the router. perhaps some of this is better with the new router. (Still using router-deprecated for this)

**Comment by geraldmorisseau** [comment]
hello, I have the same issue.
How did you corrected it ?

**Comment by zoechi** [comment]
@geraldmorisseau I saw this a few times when injecting failed because no provider was registered or because the type used for a constructor parameter was not imported.

**Comment by deebloo** [comment]
@geraldmorisseau the issue was happening to me when I was passing in undefined as a provider. for ME it was happening because of how and when my barrel was being created. Log out all of your providers to make sure none of them are undefined.

**Comment by geraldmorisseau** [comment]
thanks, i'm goiing ton investigate with these informations.

**Comment by MartinNowak** [comment]
I had this problem b/c an outdated `js` file next to the `ts` file was picked up, this broke the DI, but the error message is really not useful.
It was a bit easier to debug by using/overwriting the entry module, rather than editing the whole app.
```ts
import {ReflectiveInjector} from '@angular/core';
import {HTTP_PROVIDERS} from '@angular/http';
import {Backend} from './backend.service'

var injector = ReflectiveInjector.resolveAndCreate([HTTP_PROVIDERS, Backend]);
console.dir(injector.get(Backend));
```

**Comment by geraldmorisseau** [comment]
Hello, thanks for all your replies.
I found my problem, it's described here : https://github.com/angular/angular/issues/8586
In only one component of my application, the import wth a barrel index.ts caused this problem (I don't understand why only in this component ?).

**Comment by kokokenada** [comment]
+1 

It must be possible to improve the error message.  Huge time waste trying to figure out your error.

**Comment by rlegrand** [comment]
Hello,
I have the same issue, and doesn't seem to be for same reasons than previous poeple here.
You can reproduce it here: http://plnkr.co/edit/2GQwDiDAm3S0dGWEAIFD?p=preview

In **manage-definitions.ts**, remove the **MenuService** injection and it will work fine...
In the generated ES5 for manage-definitions, __decorage method is called before menu_service_1 is correctly set...

Seems to be related to **SystemJs** more than **angular** in my advice.
Any idea or workaround?

**Comment by ericmartinezr** [comment]
@rlegrand 
```
import {MenuService} from '../../services/menu-service';
console.log('MenuService', MenuService);
```
Put that in every file where you're importing MenuService. It's undefined. Check the paths, you're injecting something undefined in the constructor.

**Comment by pkozlowski-opensource** [comment]
Ironically I've bumped this one myself today :-/ So I'm going to re-open this issue and see if / what we can do about this error message.

**Comment by rlegrand** [comment]
@ericmartinezr 
Don't understand your answer,
in app.ts, the import works fine.
Why would it be undefined in manage-definitions?

**Comment by ericmartinezr** [comment]
@rlegrand I have no idea why

@pkozlowski-opensource this would be a dup of https://github.com/angular/angular/issues/8245, I have a repro there. Your choice which one should be open.

**Comment by pkozlowski-opensource** [comment]
Oh, right, I think we can just keep #8245. Thnx @ericmartinezr !

**Comment by edaddou** [comment]
I'm going through the same issue and I have only my app.ts file and boot.ts. I can't figure out why.

PS I'm using Laravel as backend and Gulp

**Comment by kokokenada** [comment]
This seems to mean, “you made a mistake somewhere in your app with dependency injection”.

The only effective way I had of debugging it is to do a binary search on your components.  E.g. Remove some of the components that you suspect the problem lies and rerun your app.  If the error has gone, you now know the error is somewhere in what you just removed.  Continue to search by adding or removing until you isolated the problem

**Comment by Hongbo-Miao** [comment]
In my case, when I use **angular2-infinite-scroll 0.1.3** on my local machine I got this [same error here](https://github.com/orizens/angular2-infinite-scroll/issues/28#issuecomment-226467954).

However, when I use same codes in plunker, I got a different error.

**Comment by edaddou** [comment]
It was actually a problem in the constructor, it doesn't accept any parameters, I kept in looking for similar problems but didn't figure out the problem yet.

**Comment by Hongbo-Miao** [comment]
In **angular2-infinite-scroll**, the bug was fixed by changing from:
`constructor(private element: ElementRef | any)`
to
`constructor(private element: ElementRef)`
